### PR TITLE
feat: add retireval_top_n to config in env

### LINKED ISSUE #11068: The Rerank model in RAG needs to support independent score_threshold and top_k
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

Currently, the retrieval model and the re-ranking stage share the score_threshold and top_k parameters, which currently have the following problems:
1. The retrieval model and the re-ranking model have different score distributions and need to use different thresholds, which cannot be supported under the current solution;
2. In order to ensure the integrity of the information, as much text as possible is generally recalled during the retrieval stage, and compressed to a relatively small number after the re-ranking is completed to match the input context of the large model. Currently, when top_k is shared, if top_k is set too small, the retrieval stage cannot recall enough content, affecting the recall rate, or if top_k is set too large, it exceeds the context of the large model;

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by ProseGuys**
# Summary

Solution: Add a new retrieval-related parameter `top_n` in the environment variables. During the retrieval process, use the default `top_k` parameter to obtain the `top_k` most relevant slices. When reranking, use the `top_n` parameter to return the `top_n` slices.

> [!Tip]
> Close issue syntax: `Fixes #<issue number>` or `Resolves #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.
Resolves  #11068 

# Screenshots

<table>
  <tr>
  <td>Before: </td>
  <td>After: </td>
  </tr>
  <tr>
  <td>...</td>
  <td>...</td>
  </tr>
</table>

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

File: docker/docker-compose.yaml [hunk 1]
```diff
@@ -287,6 +287,7 @@ x-shared-env: &shared-api-worker-env
   OCEANBASE_CLUSTER_NAME: ${OCEANBASE_CLUSTER_NAME:-difyai}
   OCEANBASE_MEMORY_LIMIT: ${OCEANBASE_MEMORY_LIMIT:-6G}
   CREATE_TIDB_SERVICE_JOB_ENABLED: ${CREATE_TIDB_SERVICE_JOB_ENABLED:-false}
+  RETRIEVAL_TOP_N: ${RETRIEVAL_TOP_N:-}
```
Comment (crazywoola) [inline docker/docker-compose.yaml]: Try to give a example count here to avoid potential env loading problems.

File: api/.env.example [hunk 2]
```diff
@@ -410,4 +410,6 @@ POSITION_PROVIDER_EXCLUDES=
 # Reset password token expiry minutes
 RESET_PASSWORD_TOKEN_EXPIRY_MINUTES=5
 
-CREATE_TIDB_SERVICE_JOB_ENABLED=false
\ No newline at end of file
+CREATE_TIDB_SERVICE_JOB_ENABLED=false
+
+RETRIEVAL_TOP_N=
```
Comment (crazywoola) [inline api/.env.example]: This line as well.

File: api/.env.example [hunk 3]
```diff
@@ -410,4 +410,6 @@ POSITION_PROVIDER_EXCLUDES=
 # Reset password token expiry minutes
 RESET_PASSWORD_TOKEN_EXPIRY_MINUTES=5
 
-CREATE_TIDB_SERVICE_JOB_ENABLED=false
\ No newline at end of file
+CREATE_TIDB_SERVICE_JOB_ENABLED=false
+
+RETRIEVAL_TOP_N=0
```
Comment (ProseGuys) [inline api/.env.example:415]: if RETRIEVAL_TOP_N is 0 , dataset retrieval will use original top_k params

**Comment by nadirvishun** [comment]
Currently, in my own project, I am also using similar environment variables to simply distinguish between `topK` and `topN`. However, a better approach would be to differentiate them on the knowledge base configuration interface (and also add a `threshold` parameter), rather than handling it through coding.

**Comment by nadirvishun** [comment]
It is not recommended to merge this pull because the current logic is that the `topK` configured in the knowledge base affects both `topK` and `topN` simultaneously. If merged, a default value would need to be set in the environment variable. However, not everyone understands code, not everyone comprehend the meaning of this configuration. Many people will not set it, leading to significant differences between the results after merging and before merging, which would be a break change. The best approach would be to add an additional visual configuration on the knowledge base interface.

**Comment by ProseGuys** [comment]
> Currently, in my own project, I am also using similar environment variables to simply distinguish between `topK` and `topN`. However, a better approach would be to differentiate them on the knowledge base configuration interface (and also add a `threshold` parameter), rather than handling it through coding.目前，在我自己的项目中，我也在使用类似的环境变量来简单区分`topK`和`topN` 。然而，更好的方法是在知识库配置界面上区分它们（并添加`threshold`参数），而不是通过编码来处理。

Yes, this is only a temporary solution for users with a rag background. Adding relevant parameters on the page side would involve product design, so we have not made any relevant modifications.

**Review by crazywoola — APPROVED** [review]

File: api/core/rag/datasource/retrieval_service.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 from flask import Flask, current_app
 
+from configs import DifyConfig
```
Comment (laipz8200) [inline api/core/rag/datasource/retrieval_service.py:6]: This is the wrong way to use the config object.

Comment (ProseGuys) [reply api/core/rag/datasource/retrieval_service.py:6]: Sorry, I made a stupid mistake and caused a bad impact on your, but the revised submission

Comment (laipz8200) [reply api/core/rag/datasource/retrieval_service.py:6]: @ProseGuys We reverted this change in this version and reopened the related issue.

**Pull request merged**